x/upgrade/client/cli/tx.go

```diff
@@ -58,14 +62,24 @@ func NewCmdSubmitUpgradeProposal() *cobra.Command {
 			if !noValidate {
 				prop := content.(*types.SoftwareUpgradeProposal)
 				var daemonName string
+				var noValCS, noValInfoCS, noValBinCS bool
 				if daemonName, err = cmd.Flags().GetString(FlagDaemonName); err != nil {
 					return err
 				}
+				if noValCS, err = cmd.Flags().GetBool(FlagNoValidateCS); err != nil {
+					return err
+				}
+				if noValInfoCS, err = cmd.Flags().GetBool(FlagNoValidateInfoCS); err != nil {
+					return err
+				}
+				if noValBinCS, err = cmd.Flags().GetBool(FlagNoValidateBinCS); err != nil {
+					return err
+				}
 				var planInfo *plan.Info
-				if planInfo, err = plan.ParseInfo(prop.Plan.Info); err != nil {
+				if planInfo, err = plan.ParseInfo(prop.Plan.Info, !noValCS && !noValInfoCS); err != nil {
 					return err
 				}
-				if err = planInfo.ValidateFull(daemonName); err != nil {
+				if err = planInfo.ValidateFull(daemonName, !noValCS && !noValBinCS); err != nil {
```

Comment:
passing `!noValCS && !noValInfoCS` as argument `requireChecksums`:
 is it possible that `!noValCS == true && !noValInfoCS == false` which will result skipping validate the checksum.

Reply:
It's a little tricky because of the negation and the need for flags to define `true` values, but this is correct.
- In order for `!noValCS` to be `true`, `noValCS` must be `false`. That means they did NOT provide the `--no-validate-checksums` flag.
- In order for `!noValInfoCS ` to be `false`, `noValInfoCS` must be `true`. That means they DID provided the `--no-validate-info-checksum` flag. This indicates that they are okay with a url in the `info` field not having a `checksum` query parameter.

So the result of providing `!noValCS == true && !noValInfoCS == false` => `false` as the `requireChecksum` argument to `ParseInfo` is what is desired. I.e. do not require a url in the `info` field to have a `checksum` query parameter.

Providing the `--no-validate-checksums` flag is the same as providing both `--no-validate-info-checksum` and `--no validate-binary-checksums`.

Here's a full list of scenarios:
1. Flags provided: none.
    Expected behavior: checksums are required on info field url and binary urls.
    `noValCS` is `false`, `noValInfoCS` is `false`, `noValBinCS` is `false`.
    `!noValCS && !noValInfoCS` => `true && true` => `true`
    `!noValCS && !noValBinCS` => `true && true` => `true`
    Outcome: checksums are required on info field url and binary urls.
2. Flags provided: `--no-validate-checksum`
    Expected behavior: no checksums are required.
    `noValCS` is `true`, `noValInfoCS` is `false`, `noValBinCS` is `false`.
    `!noValCS && !noValInfoCS` => `false && true` => `false`
    `!noValCS && !noValBinCS` => `false && true` => `false`
    Outcome: no checksums are required.
3. Flags provided: `--no-validate-info-checksum`
    Expected behavior: checksums are required on the binary urls, but not the info url.
    `noValCS` is `false`, `noValInfoCS` is `true`, `noValBinCS` is `false`.
    `!noValCS && !noValInfoCS` => `true && false` => `false`
    `!noValCS && !noValBinCS` => `true && true` => `true`
    Outcome: checksums are required on the binary urls, but not the info url.
4. Flags provided: `--no validate-binary-checksums`
    Expected behavior: checksum is required on info field url, but not the binary urls.
    `noValCS` is `false`, `noValInfoCS` is `false`, `noValBinCS` is `true`.
    `!noValCS && !noValInfoCS` => `true && true` => `true`
    `!noValCS && !noValBinCS` => `true && false` => `false`
    Outcome: checksums are required on info field url and binary urls.
5. Flags provided: `--no-validate-checksum` `--no-validate-info-checksum`
    Expected behavior: no checksums are required.
    `noValCS` is `true`, `noValInfoCS` is `true`, `noValBinCS` is `false`.
    `!noValCS && !noValInfoCS` => `false && false` => `false`
    `!noValCS && !noValBinCS` => `false && true` => `false`
    Outcome: no checksums are required.
6. Flags provided: `--no-validate-checksum` `--no validate-binary-checksums`
    Expected behavior: no checksums are required.
    `noValCS` is `true`, `noValInfoCS` is `false`, `noValBinCS` is `true`.
    `!noValCS && !noValInfoCS` => `false && true` => `false`
    `!noValCS && !noValBinCS` => `false && false` => `false`
    Outcome: no checksums are required.
7. Flags provided: `--no-validate-info-checksum` `--no validate-binary-checksums`
    Expected behavior: no checksums are required.
    `noValCS` is `false`, `noValInfoCS` is `true`, `noValBinCS` is `true`.
    `!noValCS && !noValInfoCS` => `true && false` => `false`
    `!noValCS && !noValBinCS` => `true && false` => `false`
    Outcome: no checksums are required.
8. Flags provided: `--no-validate-checksum` `--no-validate-info-checksum` `--no validate-binary-checksums`
    Expected behavior: no checksums are required.
    `noValCS` is `true`, `noValInfoCS` is `true`, `noValBinCS` is `true`.
    `!noValCS && !noValInfoCS` => `false && false` => `false`
    `!noValCS && !noValBinCS` => `false && false` => `false`
    Outcome: no checksums are required.